crypto/keys/ethsecp256k1/ethsecp256k1.go

```diff
@@ -0,0 +1,202 @@
+
+	tmcrypto "github.com/tendermint/tendermint/crypto"
+)
+
+const (
+	// PrivKeySize defines the size of the PrivKey bytes
+	PrivKeySize = 32
+	// PubKeySize defines the size of the PubKey bytes
+	PubKeySize = 33
+	// KeyType is the string constant for the EthSecp256k1 algorithm
+	KeyType = "ethsecp256k1"
+)
+
+// Amino encoding names
+const (
+	// PrivKeyName defines the amino encoding name for the EthSecp256k1 private key
+	PrivKeyName = "tendermint/PrivKeyEthSecp256k1"
+	// PubKeyName defines the amino encoding name for the EthSecp256k1 public key
+	PubKeyName = "tendermint/PubKeyEthSecp256k1"
+)
+
+// ----------------------------------------------------------------------------
+// secp256k1 Private Key
+
+var (
+	_ cryptotypes.PrivKey  = &PrivKey{}
+	_ codec.AminoMarshaler = &PrivKey{}
+)
+
+// GenerateKey generates a new random private key. It returns an error upon
+// failure.
+func GenerateKey() (*PrivKey, error) {
+	priv, err := ethcrypto.GenerateKey()
+	if err != nil {
+		return nil, err
+	}
+
+	return &PrivKey{
+		Key: ethcrypto.FromECDSA(priv),
+	}, nil
+}
+
+// Bytes returns the byte representation of the ECDSA Private Key.
+func (privKey PrivKey) Bytes() []byte {
+	return privKey.Key
+}
+
+// PubKey returns the ECDSA private key's public key.
+func (privKey PrivKey) PubKey() cryptotypes.PubKey {
+	ecdsaPrivKey := privKey.ToECDSA()
+	return &PubKey{
+		Key: ethcrypto.CompressPubkey(&ecdsaPrivKey.PublicKey),
+	}
+}
+
+// Equals returns true if two ECDSA private keys are equal and false otherwise.
+func (privKey PrivKey) Equals(other cryptotypes.LedgerPrivKey) bool {
+	return privKey.Type() == other.Type() && subtle.ConstantTimeCompare(privKey.Bytes(), other.Bytes()) == 1
+}
+
+// Type returns ethsecp256k1
+func (privKey PrivKey) Type() string {
+	return KeyType
+}
+
+// MarshalAmino overrides Amino binary marshalling.
+func (privKey PrivKey) MarshalAmino() ([]byte, error) {
+	return privKey.Key, nil
+}
+
+// UnmarshalAmino overrides Amino binary marshalling.
+func (privKey *PrivKey) UnmarshalAmino(bz []byte) error {
+	if len(bz) != PrivKeySize {
+		return fmt.Errorf("invalid privkey size, expected %d got %d", PrivKeySize, len(bz))
+	}
+	privKey.Key = bz
+
+	return nil
+}
+
+// MarshalAminoJSON overrides Amino JSON marshalling.
+func (privKey PrivKey) MarshalAminoJSON() ([]byte, error) {
+	// When we marshal to Amino JSON, we don't marshal the "key" field itself,
+	// just its contents (i.e. the key bytes).
+	return privKey.MarshalAmino()
+}
+
+// UnmarshalAminoJSON overrides Amino JSON marshalling.
+func (privKey *PrivKey) UnmarshalAminoJSON(bz []byte) error {
+	return privKey.UnmarshalAmino(bz)
+}
+
+// Sign creates a recoverable ECDSA signature on the secp256k1 curve over the
+// provided hash of the message. The produced signature is 65 bytes
+// where the last byte contains the recovery ID.
+func (privKey PrivKey) Sign(msg []byte) ([]byte, error) {
+	return ethcrypto.Sign(ethcrypto.Keccak256Hash(msg).Bytes(), privKey.ToECDSA())
```

Comment:
Maybe we shouldn't actually hash the message, prior to signing. The ethereum `Signer.Hash` transaction already hashed the contents of the `Transaction` so this implementation would hash the contents twice

Reply:
At least that was the purpose of the `if` condition. To be able to sign the transaction without exposing the private key from the keyring (and just use the keyring `Signer` interface)